k2/csrc/top_sort.cu

```diff
@@ -67,14 +67,19 @@ FsaVec RenumberFsaVec(FsaVec &fsas, const Array1<int32_t> &order,
   int32_t ans_num_arcs = ans_shape.NumElements();
   Array1<Arc> ans_arcs(c, ans_num_arcs);
   int32_t *arc_map_data;
-  if (arc_map) {
+  if (arc_map != nullptr) {
     *arc_map = Array1<int32_t>(c, ans_num_arcs);
     arc_map_data = arc_map->Data();
   } else {
     arc_map_data = nullptr;
   }
+
   const Arc *fsas_arcs = fsas.values.Data();
   Arc *ans_arcs_data = ans_arcs.Data();
+  Renumbering arcs_renumbering(c, ans_num_arcs);
+  // we will remove those arcs whose dest_state is not in `order`,
```

Comment:
@qindazhu as the algorithm is written, it would be an error if any of the input states were not present in the output, i.e. it would indicate cycles in the input or that the start-state had arcs entering it from a non-start state.  
Therefore I am concerned to see you spending a bunch of effort to deal with this situation.   We can just return the empty FsaVec.  We need to clarify that a result whose size differs from that of the input is an error; we can just always return the empty FsaVec in this situation.

Reply:
I'm confused here, the documentation said 
https://github.com/k2-fsa/k2/blob/a377250db4b232bd8801c86a2973939325057d6d/k2/csrc/fsa_utils.h#L110
Suppose we only have one Fsa in `fsas`, and the order is `[0 2 3]`, there is an arc [0, 1, x ,x], but as state `1` is not present in the order, don't we need remove this arc? Did I misunderstand the documentation?

Reply:
Ah OK, this is RenumberFsaVec().

Reply:
@qindazhu couldn't we just say it is an error if there are any arcs from states that are kept to those that are not kept, for now, though?  I don't see us ever needing this in the situation where some of the non-kept states have arcs going to them; normally when we do things like pruning we'll do that after top-sorting anyway and we'll do the renumbering ourselves.

Reply:
I think it would be better if when you find bugs that are not trivial, you ask me about it.  I don't tend to write code that needs fundamental rewrites, just code with small bugs.

Reply:
OK. And just wondering why not we just require that the orders.size must be equal to fas.TotSize(1), I just feel it's hard for users to remove those arcs with dest-state not kept, I don't see the situation.

Reply:
I suppose I wanted it as general as possible.  How about for now we don't require that order.Dim() == fsas.TotSize(1), but do say that it's an error if there are arcs from kept to not-kept states.  We can relax that restriction later if we ever have a case where that can happen.

Reply:
OK, will do (and also update the documentation)